include/wally_psbt.h

```diff
@@ -0,0 +1,430 @@
+struct keypath_item {
+    unsigned char pubkey[65];
+    struct key_origin_info origin;
+};
+
+/** A map of public keys to BIP 32 fingerprint and derivation paths */
+struct keypath_map {
+    struct keypath_item *items;
+    size_t num_items;
+    size_t items_allocation_len;
+};
+
+/** Item in partial signatures map */
+struct partial_sigs_item {
+    unsigned char pubkey[65];
+    unsigned char *sig;
+    size_t sig_len;
+};
+
+/** A map of public key's to signatures */
+struct partial_sigs_map {
+    struct partial_sigs_item *items;
+    size_t num_items;
+    size_t items_allocation_len;
+};
+
+/** Unknown item */
+struct unknowns_item {
+    unsigned char *key;
+    size_t key_len;
+    unsigned char *value;
+    size_t value_len;
+};
+
+/** Unknown items map */
+struct unknowns_map {
+    struct unknowns_item *items;
+    size_t num_items;
+    size_t items_allocation_len;
+};
+
+/** A psbt input map */
+struct wally_psbt_input {
+    struct wally_tx *non_witness_utxo;
+    struct wally_tx_output *witness_utxo;
+    unsigned char *redeem_script;
+    size_t redeem_script_len;
+    unsigned char *witness_script;
+    size_t witness_script_len;
+    unsigned char *final_script_sig;
+    size_t final_script_sig_len;
+    struct wally_tx_witness_stack *final_witness;
+    struct keypath_map *keypaths;
+    struct partial_sigs_map *partial_sigs;
+    struct unknowns_map *unknowns;
+    uint32_t sighash_type;
+};
+
+/** A psbt output map */
+struct wally_psbt_output {
+    unsigned char *redeem_script;
+    size_t redeem_script_len;
+    unsigned char *witness_script;
+    size_t witness_script_len;
+    struct keypath_map *keypaths;
+    struct unknowns_map *unknowns;
+};
+
+/** A parsed bitcoin transaction */
+struct wally_psbt {
+    struct wally_tx *tx;
+    struct wally_psbt_input *inputs;
+    size_t num_inputs;
+    size_t inputs_allocation_len;
+    struct wally_psbt_output *outputs;
+    size_t num_outputs;
+    size_t outputs_allocation_len;
+    struct unknowns_map *unknowns;
+};
+#endif /* SWIG */
+
+/**
+ * Allocate and initialize a new keypath map.
+ *
+ * :param alloc_len: The number of items to allocate.
+ * :param output: Destination for the new keypath map
+ */
+WALLY_CORE_API int keypath_map_init_alloc(size_t alloc_len, struct keypath_map **output);
+
+#ifndef SWIG_PYTHON
+/**
+ * Free a keypath map allocated by `keypath_map_init_alloc`.
+ *
+ * :param keypaths: The keypath map to free.
+ */
+WALLY_CORE_API int keypath_map_free(struct keypath_map *keypaths);
+#endif /* SWIG_PYTHON */
+
+/**
+ * Add an item to a keypath map
+ *
+ * :param keypaths: The keypath map to add to
+ * :param pubkey: The pubkey to add
+ * :param fingerprint: The master key fingerprint for the pubkey
+ * :param path: The BIP32 derivation path for the pubkey
+ * :param path_len: The number of items in path
+ */
+WALLY_CORE_API int add_new_keypath(struct keypath_map *keypaths,
+                                   unsigned char pubkey[65],
```

Comment:
Could these instances of literal 65 be EC_PUBLIC_KEY_UNCOMPRESSED_LEN?

Reply:
Done